### PR TITLE
Add new plugin to update fpc fp devices

### PR DESCRIPTION
**Pull request by haoweilo**
Type of pull request:

- [x] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/fpc/fu-fpc-device.c [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+		     guint8 *data,
+		     gsize length,
+		     gboolean device2host,
+		     GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
+	gsize actual_len = 0;
+	gboolean ret;
+
+	if ((data == NULL) && (length > 0)) {
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:115]: No need for the inner brackets.

File: plugins/fpc/fu-fpc-device.c [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+	g_autoptr(GPtrArray) intfs = NULL;
+
+	intfs = g_usb_device_get_interfaces(usb_device, error);
+	if (intfs == NULL)
+		return FALSE;
+	for (guint i = 0; i < intfs->len; i++) {
+		GUsbInterface *intf = g_ptr_array_index(intfs, i);
+		if (g_usb_interface_get_class(intf) == FPC_DEVICE_DFU_MODE_CLASS &&
+		    g_usb_interface_get_protocol(intf) == FPC_DEVICE_DFU_MODE_PORT) {
+			g_debug("device run in bootloader");
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:162]: the g_debug isn't needed

File: plugins/fpc/fu-fpc-device.c [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+				   (guint8 *)dfu_status,
+				   sizeof(FuFpcDfuStatus),
+				   TRUE,
+				   FALSE,
+				   error)) {
+		g_prefix_error(error, "failed to get status: ");
+		return FALSE;
+	}
+
+	if ((dfu_status->bstatus != 0) || (dfu_status->bstate != FPC_DFU_IDLE)) {
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:263]: no need for inner brackets

File: plugins/fpc/fu-fpc-device.c [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+
+	if (!fu_device_retry_full(FU_DEVICE(self),
+				  fu_fpc_device_check_dfu_status,
+				  FPC_DFU_MAX_ATTEMPTS,
+				  20,
+				  (gpointer)&dfu_status,
+				  error))
+		return FALSE;
+
+	if (dfu_status.bmax_payload_size)
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:302]: add `> 0` please

File: plugins/fpc/fu-fpc-device.c [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+fu_fpc_device_setup(FuDevice *device, GError **error)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	/* setup */
+	if (!FU_DEVICE_CLASS(fu_fpc_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	self->is_moc_device = (fu_usb_device_get_pid(FU_USB_DEVICE(device)) >> 2 & 0x1) ? TRUE
+											: FALSE;
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:361]: is this **always** going to be true, or would a per-device quirk be a better idea?

File: plugins/fpc/fu-fpc-device.c [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+	FuFpcDfuStatus dfu_status = {0};
+	gboolean ret_dnload = TRUE;
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:394]: if the values are the daemon-provided ones then you can remove this to get rid of the red warning.

File: plugins/fpc/fu-fpc-device.c [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+
+		if (!fu_fpc_device_dfu_cmd(self,
+					   FPC_CMD_DFU_DNLOAD,
+					   (guint16)i,
+					   req->data,
+					   (gsize)req->len,
+					   FALSE,
+					   FALSE,
+					   &error_local)) {
+			ret_dnload = FALSE;
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:435]: I think just g_set_error() and return FALSE here, no ret_dnload is needed.

File: plugins/fpc/fu-fpc-device.c [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+			break;
+		}
+
+		if (!fu_device_retry_full(device,
+					  fu_fpc_device_check_dfu_status,
+					  FPC_DFU_MAX_ATTEMPTS,
+					  20,
+					  (gpointer)&dfu_status,
+					  &error_local)) {
+			ret_dnload = FALSE;
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:445]: I think just g_set_error() and return FALSE here, no ret_dnload is needed.

File: plugins/fpc/fu-fpc-device.c [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+			break;
+		}
+
+		/* update progress */
+		fu_progress_set_percentage_full(fu_progress_get_child(progress),
+						(gsize)i + 1,
+						(gsize)chunks->len);
+	}
+
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:454]: i think you probably want to `fu_progress_step_done(progress);` here and add another `fu_progress_add_step` for the checking DFU status.

File: plugins/fpc/fu-fpc-device.c [hunk 10]
```diff
@@ -0,0 +1,583 @@
...
+					    value,
+					    0x0000,
+					    data,
+					    length,
+					    length ? &actual_len : NULL,
+					    FPC_USB_TRANSFER_TIMEOUT,
+					    NULL,
+					    error);
+	if (!ret)
+		return FALSE;
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:112]: remove the temp `ret` variable please, just do a big `if (!g_usb_device_control_transfer`

File: plugins/fpc/fu-fpc-device.c [hunk 11]
```diff
@@ -0,0 +1,583 @@
...
+{
+	guint32 version = 0;
+	gsize data_len = 0;
+	FuEndianType endian_type = G_LITTLE_ENDIAN;
+	g_autofree guint8 *data = NULL;
+	g_autofree gchar *str_version = NULL;
+
+	if (fu_device_has_private_flag(FU_DEVICE(self), FU_FPC_DEVICE_FLAG_RTS_DEVICE)) {
+		endian_type = G_BIG_ENDIAN;
+	}
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:220]: no need for `{}` here

File: plugins/fpc/fu-fpc-device.c [hunk 12]
```diff
@@ -0,0 +1,583 @@
...
+
+static void
+fu_fpc_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuFpcDevice *self = FU_FPC_DEVICE(device);
+
+	fu_string_append_kx(str, idt, "Max block size", self->max_block_size);
+	fu_string_append_kb(str,
+			    idt,
+			    "Legacy DFU",
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:352]: This should be CamelCase, so in this case `LegacyDfu`

File: plugins/fpc/fu-fpc-device.c [hunk 13]
```diff
@@ -0,0 +1,583 @@
...
+			    fu_device_has_private_flag(device, FU_FPC_DEVICE_FLAG_LEGACY_DFU));
+	fu_string_append_kb(str,
+			    idt,
+			    "Is a moc device",
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:356]: `MocDevice`

File: plugins/fpc/fu-fpc-device.c [hunk 14]
```diff
@@ -0,0 +1,583 @@
...
+	fu_device_add_protocol(FU_DEVICE(self), "com.fingerprints.dfupc");
+	fu_device_set_summary(FU_DEVICE(self), "FPC fingerprint sensor");
+	fu_device_set_install_duration(FU_DEVICE(self), 15);
+	fu_device_set_firmware_size_min(FU_DEVICE(self), 0x10000);
+	fu_device_set_firmware_size_max(FU_DEVICE(self), 0x50000);
+	fu_usb_device_add_interface(FU_USB_DEVICE(self), FPC_USB_INTERFACE);
+	fu_device_register_private_flag(FU_DEVICE(self),
+					FU_FPC_DEVICE_FLAG_MOH_DEVICE,
+					"moh-device");
+	fu_device_register_private_flag(FU_DEVICE(self), FU_FPC_DEVICE_FLAG_RTS_DEVICE, "rts");
```
Comment (hughsie) [inline plugins/fpc/fu-fpc-device.c:566]: is `moh-device` and `rts` mutually exclusive? i.e., can you have a device that's bot moh **and** rts?

Comment (haoweilo) [reply plugins/fpc/fu-fpc-device.c:561]: Yes, it's mutually exclusive. Not all moh devices are rts. :)

**Comment by hughsie** [comment]
I've fixed the CI failure, and pushed this to the main branch -- thanks!

**Comment by haoweilo** [comment]
Thanks a lot for the help.

**Pull request closed without merging**